src/Products/ZCatalog/Catalog.py

```diff
@@ -845,27 +845,16 @@ def _sort_nbest(self, actual_result_count, result, rs,
                     worst = keys[0]
             result.reverse()
         else:
-            for did in rs:
-                try:
-                    key = index_key_map[did]
-                    full_key = (key, )
-                    for km in second_indexes_key_map:
-                        full_key += (km[did], )
-                except KeyError:
-                    # This document is not in the sort key index, skip it.
-                    actual_result_count -= 1
-                else:
-                    if n >= limit and key <= worst:
-                        continue
-                    i = bisect(keys, key)
-                    keys.insert(i, key)
-                    result.insert(i, (full_key, did, self.__getitem__))
-                    if n == limit:
-                        del keys[0], result[0]
-                    else:
-                        n += 1
-                    worst = keys[0]
-            result = multisort(result, sort_spec)
+            # we have multi index sorting
+            result = self._multi_index_nbest(
+                actual_result_count,
```

Comment:
Moving the sorting code into a separate method (which is in general a good idea) prevents `actual_result_count` to be updated. You must either wrap (`int`) `actual_result_count` in a "mutable" object (i.e. updatable in place) or return the updated value as part of the return value (and reassign). `ZCatalog` filters out hits for which at least one of the sort indexes lacks a value - a test examining the correct `actual_result_count` thus would ensure that some hits lack a sort value and verify the correct result size.